### DISCUSSION TITLE
Per-flow source port?

### DISCUSSION THREAD

**Discussion by jlearman** [strongswan/strongswan#1891] [Q&A]
When we did a Proof of Concept for using Strongswan, configured using etc/ipsec.conf, we saw that the data traffic for a single SA used different source ports, presumably based on the original packet's flow hash, as recommended by the IPsec RFC.  This was great, as we want cloud per-flow limiting to apply to actual flows rather than per IPsec tunnel.

We didn't configure anything special to get this behavior.  We're using NAT-T, so the port is 4500.  It's possible we didn't have NAT-T enabled for the PoC.

We're now configuring Strongswan via `govici`.  I noticed that the source port is now always 4500.  Is there a way to enable this feature (to set source port based on flow hash, or some reasonable facsilime)?

**Comment by tobiasbrunner** [comment]
> we saw that the data traffic for a single SA used different source ports, presumably based on the original packet's flow hash, as recommended by the IPsec RFC.

Neither does strongSwan or the Linux kernel have (or ever had) this feature, nor does any IPsec RFC I'm aware of state anything the like.

> This was great, as we want cloud per-flow limiting to apply to actual flows rather than per IPsec tunnel.

Not sure what mechanism you are referring to exactly but that seems to be problematic in regards to replay protection.

> It's possible we didn't have NAT-T enabled for the PoC.

Hm, only with NAT-T i.e. UDP-encapsulation will there even be UDP ports that could have different source ports.

**Comment by jlearman** [comment]
Sorry, I goofed.  We got varying source port (based on original incoming packet's flow hash) using IPsec/GENEVE tunnels.  I completely forgot that this happened only when running over GENEVE tunnels.  (And it's the GENEVE spec that specifies that the source port should be a function of the flow 5-tuple, not any IPsec spec.)

Thanks for the help!
